File: frontend/src/services/projectService.js
// frontend/src/services/projectService.js
import api from "@/api/axios";

/* -------------------------------------------------------------------------- */
/*  PROJECTS – CORE CRUD                                                      */
/* -------------------------------------------------------------------------- */

// Get all projects the current user can see (my + assigned via tasks)
export async function getMyProjects() {
  const { data } = await api.get("/projects/my");
  return data.map(p => ({
    ...p,
    deadline: p.deadline ? new Date(p.deadline) : null
  }));
}

// Admin: Get every project in the system
export async function getAllProjects() {
  const { data } = await api.get("/projects");
  return data;
}

// Get a single project by ID
export async function getProjectById(projectId) {
  const { data } = await api.get(`/projects/${projectId}`);
  return data;
}

// Create a new project (evaluator only)
export async function createProject(payload) {
  const { data } = await api.post("/projects", payload);
  return {
    ...data,
    deadline: data.deadline ? new Date(data.deadline) : null,
  };
}

// Update an existing project (evaluator only)
export async function updateProject(projectId, payload) {
  const { data } = await api.put(`/projects/${projectId}`, payload);
  return {
    ...data,
    deadline: data.deadline ? new Date(data.deadline) : null,
  };
}

// Delete a project (evaluator only)
export async function deleteProject(projectId) {
  await api.delete(`/projects/${projectId}`);
}

/* -------------------------------------------------------------------------- */
/*  ADMIN – USER-PROJECT RELATIONSHIP                                         */
/* -------------------------------------------------------------------------- */

// Admin: Get all projects a specific user is involved in
export async function getUserProjects(userId) {
  const { data } = await api.get(`/projects/user/${userId}`);
  return data;
}

/* -------------------------------------------------------------------------- */
/*  EMPLOYEES – FETCHING & ASSIGNMENT                                         */
/* -------------------------------------------------------------------------- */

// Get all employees (for assignment dropdown)
export async function getAllEmployees() {
  const { data } = await api.get("/users?role=Employee");
  return data;
}

// Get employee IDs currently assigned to a project (preferred endpoint)
export async function getAssignedEmployees(projectId) {
  try {
    const { data } = await api.get(`/projects/${projectId}/assignments`);
    return data.map(a => a.employeeId || a.userId);
  } catch {
    // Fallback: derive from tasks if endpoint not available
    return getEmployeesByProject(projectId);
  }
}

// Legacy fallback: extract employees from project tasks
export async function getEmployeesByProject(projectId) {
  const project = await getProjectById(projectId);
  const tasks = project.tasks || [];
  return tasks
    .filter(t => t.assignedTo)
    .map(t => ({
      id: t.assignedTo.id,
      name: t.assignedTo.name,
      email: t.assignedTo.email,
      taskTitle: t.title,
      status: t.status,
    }));
}

// Assign/replace employees on a project (evaluator/admin)
export async function assignEmployeesToProject(projectId, employeeIds) {
  await api.post(`/projects/${projectId}/assign`, { employeeIds });
}

/* -------------------------------------------------------------------------- */
/*  PROJECT COMPLETION                                                        */
/* -------------------------------------------------------------------------- */

// Mark a project as completed (evaluator only)
export const markProjectAsCompleted = async (projectId) => {
  const { data } = await api.patch(`/projects/${projectId}/complete`);
  return data;
};

/* -------------------------------------------------------------------------- */
/*  DEPRECATED / LEGACY                                                       */
/* -------------------------------------------------------------------------- */

// Deprecated: use getMyProjects() instead
export async function getEvaluatorProjects() {
  console.warn("getEvaluatorProjects() is deprecated – use getMyProjects()");
  return getMyProjects();
}

File: backend/src/Services/Projects/ProjectService.cs
using Microsoft.EntityFrameworkCore;
using task_manager_api.Data;
using task_manager_api.Models;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using task_manager_api.DTOs.Users;
using task_manager_api.DTOs.Projects;

namespace task_manager_api.Services.Projects
{
    public class ProjectService : IProjectService
    {
        private readonly ApplicationDbContext _db;

        public ProjectService(ApplicationDbContext db)
        {
            _db = db;
        }

        // Get all projects with evaluator and task details
        public async Task<IEnumerable<Project>> GetAllProjectsAsync()
        {
            return await _db.Projects
                .Include(p => p.Evaluator)
                .Include(p => p.Tasks)
                .ThenInclude(t => t.AssignedTo)
                .ToListAsync();
        }

        // Get a single project by ID with full details
        public async Task<Project?> GetProjectByIdAsync(Guid id)
        {
            return await _db.Projects
                .Include(p => p.Evaluator)
                .Include(p => p.Tasks)
                .ThenInclude(t => t.AssignedTo)
                .FirstOrDefaultAsync(p => p.Id == id);
        }

        // Create a new project for the given evaluator
        public async Task<Project> CreateProjectAsync(string name, string? description, Guid evaluatorId, DateTime? deadline = null)
        {
            var project = new Project
            {
                Name = name,
                Description = description,
                EvaluatorId = evaluatorId,
                Deadline = deadline
            };

            _db.Projects.Add(project);
            await _db.SaveChangesAsync();
            return project;
        }

        // Update project details (name, description, deadline) — evaluator only
        public async Task<Project?> UpdateProjectAsync(Guid id, string? name, string? description, Guid evaluatorId, DateTime? deadline = null)
        {
            var project = await _db.Projects.FindAsync(id);
            if (project == null || project.EvaluatorId != evaluatorId)
                return null;

            if (name != null) project.Name = name;
            if (description != null) project.Description = description;
            if (deadline.HasValue) project.Deadline = deadline;

            await _db.SaveChangesAsync();
            return project;
        }

        // Delete a project — evaluator only
        public async Task<bool> DeleteProjectAsync(Guid id, Guid evaluatorId)
        {
            var project = await _db.Projects.FindAsync(id);
            if (project == null || project.EvaluatorId != evaluatorId)
                return false;

            _db.Projects.Remove(project);
            await _db.SaveChangesAsync();
            return true;
        }

        // Get all projects the user is involved in (as evaluator or assigned employee)
        public async Task<IEnumerable<Project>> GetUserProjectsAsync(Guid userId)
        {
            return await _db.Projects
                .Where(p => p.EvaluatorId == userId || p.Tasks.Any(t => t.AssignedToId == userId))
                .Include(p => p.Evaluator)
                .Include(p => p.Tasks)
                    .ThenInclude(t => t.AssignedTo)
                .ToListAsync();
        }

        // Get all projects owned by a specific evaluator
        public async Task<IEnumerable<Project>> GetProjectsByEvaluatorAsync(Guid evaluatorId)
        {
            return await _db.Projects
                .Where(p => p.EvaluatorId == evaluatorId)
                .Include(p => p.Evaluator)
                .Include(p => p.Tasks)
                .ThenInclude(t => t.AssignedTo)
                .ToListAsync();
        }

        // Assign/replace employees to a project (admin or project evaluator only)
        public async Task<bool> AssignEmployeesAsync(Guid projectId, List<Guid> employeeIds, Guid currentUserId)
        {
            var project = await _db.Projects
                .Include(p => p.AssignedEmployees)
                .FirstOrDefaultAsync(p => p.Id == projectId);

            if (project == null)
                return false;

            var currentUser = await _db.Users.FindAsync(currentUserId);
            var isAdmin = currentUser?.Role == Role.Admin;

            if (!isAdmin && project.EvaluatorId != currentUserId)
                return false;

            _db.ProjectAssignments.RemoveRange(project.AssignedEmployees);

            foreach (var empId in employeeIds.Distinct())
            {
                var employee = await _db.Users.FindAsync(empId);
                if (employee == null || employee.Role != Role.Employee)
                    continue;

                _db.ProjectAssignments.Add(new ProjectAssignment
                {
                    ProjectId = projectId,
                    UserId = empId
                });
            }

            await _db.SaveChangesAsync();
            return true;
        }

        // Get all employees currently assigned to a project
        public async Task<IEnumerable<EmployeeDto>> GetAssignedEmployeesAsync(Guid projectId)
        {
            var employees = await _db.ProjectAssignments
                .Where(a => a.ProjectId == projectId)
                .Include(a => a.User)
                .Select(a => new EmployeeDto(
                    a.User.Id,
                    a.User.Name,
                    a.User.Email))
                .ToListAsync();

            return employees;
        }

        // Get all projects a specific user is involved in (admin view)
        public async Task<List<Project>> GetProjectsForUserAsync(Guid userId)
        {
            return await _db.Projects
                .Include(p => p.Tasks)
                .Where(p =>
                    p.EvaluatorId == userId ||
                    p.Tasks.Any(t => t.AssignedToId == userId))
                .ToListAsync();
        }

        // Get all tasks in a specific project (with assignee)
        public async Task<List<TaskItem>> GetTasksForProjectAsync(Guid projectId)
        {
            return await _db.Tasks
                .Include(t => t.AssignedTo)
                .Where(t => t.ProjectId == projectId)
                .ToListAsync();
        }

        // Mark a project as completed (evaluator only, idempotent)
        public async Task<bool> CompleteProjectAsync(Guid projectId, Guid evaluatorId)
        {
            var project = await _db.Projects
                .FirstOrDefaultAsync(p => p.Id == projectId && p.EvaluatorId == evaluatorId);

            if (project == null)
                return false;

            if (project.Status == "Completed")
                return true;

            project.Status = "Completed";
            await _db.SaveChangesAsync();
            return true;
        }
    }
}